File: artifactory_generator/read_receipts.py
import re
from stitch.artifactory_generator.SimpleArtifactoryFinder import SimpleArtifactoryFinder

from artifactory_generator.smali import CLASS_RE


class ReadReceiptsFinder(SimpleArtifactoryFinder):
    """Finds what the read-receipt hook needs to pick the self variant of a receipt.

    WhatsApp does not choose between "send a receipt" and "send nothing". It
    always sends a ``receipt`` stanza and chooses its *type*: ``read`` goes to
    the sender and turns the ticks blue, ``read-self`` goes only to your own
    linked devices so they can mark the chat read. Voice notes work the same
    way with ``played`` and ``played-self``.

    The read decision is made in one class, which its own log strings call
    ``ReadReceiptUtils``: a ``(Jid, boolean)String`` method returning ``"read"``
    or ``"read-self"``. The class is anchored on ``ReadReceiptUtils/``, the
    prefix of its own log messages -- the one name here obfuscation cannot
    touch. It matched exactly one class in every build examined, while the
    class itself was renamed in all five: ``X.0gN``, ``X.0lo``, ``X.1AB``,
    ``X.1Cx``, ``X.19A``.

    The receipt-type method is pinned as the only ``public`` one whose
    signature is ``(L...;Z)Ljava/lang/String;`` and whose body holds both the
    ``"read"`` and ``"read-self"`` literals. Those literals are wire-format
    values, so they cannot be renamed either.

    The played decision has no method of its own. ``SendPlayedReceiptJobV2``
    asks the same class a ``(Jid)Z`` question, and that predicate is *shared*:
    on 2.26.36.71 it has twelve callers, of which the job is one. The rest
    include ``MessageStatusUpdateReceiptFactory``, which consults it on the
    **receiving** side and, when it answers no, rewrites an incoming ``read``
    or ``played`` status to ``delivered`` -- WhatsApp's own reciprocity rule,
    "receipts off means you do not see theirs". Forcing that predicate false
    therefore hides other people's blue ticks from the user, which is the one
    thing this feature is supposed not to do.

    So the gate is resolved *and* so is the job that may see a forced answer,
    letting the hook scope its lie to that one call site. The gate is the only
    ``public`` ``(L...;)Z`` method calling at least two of the class's own
    private ``(L...;)Z`` helpers -- the pair the receipt-type method calls,
    which is how they are identified. Deriving them from an already-pinned
    method keeps every name in this finder an output.

    The job is anchored on its class name, which lives in an unobfuscated
    package, and its method pinned as the only ``public ()V`` holding both the
    ``"played"`` wire literal and the ``PlayedSelfReceiptStore/`` log prefix.
    That method is renamed between builds like everything else (``A0F`` on
    2.26.17.72, ``A0G`` after), and on every build examined it is the one that
    calls the gate this finder resolved.

    The same finder also resolves how to read a chat's Jid as a string, which
    the per-chat lists are keyed on. ``Jid.toString()`` is not it: it returns
    ``getObfuscatedString()``, whose name says exactly what it is meant to
    become even though today it happens to delegate to the raw value. The
    accessor wanted is the one that builds the string from the ``user`` field
    and the server, found by that shape rather than by name.

    Verified to resolve all four, and nothing else, on 2.26.17.72, 2.26.27.85,
    2.26.29.74, 2.26.33.76 and 2.26.36.71.
    """

    ANCHOR = 'ReadReceiptUtils/'
    JID_CLASS = 'com/whatsapp/infra/core/jid/Jid'
    PLAYED_JOB_CLASS = 'com/whatsapp/messaging/receipts/jobqueue/job/SendPlayedReceiptJobV2'

    METHOD_RE = re.compile(
        r'^\.method (?P<modifiers>[\w ]*?)(?P<name>[\w$]+)(?P<sig>\([^)\n]*\)[\w/$;\[]+)\n'
        r'(?P<body>.*?)^\.end method$',
        re.MULTILINE | re.DOTALL)
    RECEIPT_TYPE_SIG_RE = re.compile(r'^\(L[\w/$]+;Z\)Ljava/lang/String;$')
    PLAYED_GATE_SIG_RE = re.compile(r'^\(L[\w/$]+;\)Z$')
    READ_RE = re.compile(r'const-string(?:/jumbo)? [vp]\d+, "read"')
    READ_SELF_RE = re.compile(r'const-string(?:/jumbo)? [vp]\d+, "read-self"')

    PLAYED_JOB_SIG = '()V'
    PLAYED_RE = re.compile(r'const-string(?:/jumbo)? [vp]\d+, "played"')
    PLAYED_SELF_STORE_RE = re.compile(r'const-string(?:/jumbo)? [vp]\d+, "PlayedSelfReceiptStore/')

    RAW_STRING_SIG = '()Ljava/lang/String;'
    USER_FIELD_RE = re.compile(r'->user:Ljava/lang/String;')
    GET_SERVER_RE = re.compile(r'->getServer\(\)Ljava/lang/String;')

    MIN_SHARED_HELPERS = 2

    def __init__(self, args):
        super().__init__(args)
        self.is_once = True
        self.is_found = False
        self._found_utils = False
        self._found_jid = False
        self._found_job = False

    @staticmethod
    def _class_name(class_data: str):
        match = CLASS_RE.search(class_data)
        return match.groupdict().get('name') if match is not None else None

    def class_filter(self, class_data: str) -> bool:
        return self.ANCHOR in class_data or \
            self._class_name(class_data) in (self.JID_CLASS, self.PLAYED_JOB_CLASS)

    @staticmethod
    def _calls(body: str, name: str) -> bool:
        return re.search(r'->%s\(L[\w/$]+;\)Z' % re.escape(name), body) is not None

    def extract_artifacts(self, artifacts: dict, class_data: str) -> None:
        name = self._class_name(class_data)
        if name == self.JID_CLASS:
            self._extract_jid(artifacts, class_data)
        elif name == self.PLAYED_JOB_CLASS:
            self._extract_played_job(artifacts, class_data)
        else:
            self._extract_receipt_methods(artifacts, class_data)
        self.is_found = self._found_utils and self._found_jid and self._found_job

    def _extract_jid(self, artifacts: dict, class_data: str) -> None:
        accessors = [method for method in self.METHOD_RE.finditer(class_data)
                     if 'public' in method.group('modifiers')
                     and method.group('sig') == self.RAW_STRING_SIG
                     and self.USER_FIELD_RE.search(method.group('body'))
                     and self.GET_SERVER_RE.search(method.group('body'))]
        if len(accessors) != 1:
            return
        artifacts['JID_CLASS_NAME'] = self.JID_CLASS.replace('/', '.')
        artifacts['JID_RAW_STRING_METHOD_NAME'] = accessors[0].group('name')
        self._found_jid = True

    def _extract_played_job(self, artifacts: dict, class_data: str) -> None:
        # Both anchors survive obfuscation: "played" is the wire value the job
        # sends, and the other is the prefix of the log line it writes when it
        # stores the self variant instead.
        runs = [method for method in self.METHOD_RE.finditer(class_data)
                if 'public' in method.group('modifiers')
                and method.group('sig') == self.PLAYED_JOB_SIG
                and self.PLAYED_RE.search(method.group('body'))
                and self.PLAYED_SELF_STORE_RE.search(method.group('body'))]
        if len(runs) != 1:
            return
        artifacts['PLAYED_RECEIPT_JOB_CLASS_NAME'] = self.PLAYED_JOB_CLASS.replace('/', '.')
        artifacts['PLAYED_RECEIPT_JOB_METHOD_NAME'] = runs[0].group('name')
        artifacts['PLAYED_RECEIPT_JOB_METHOD_SIG'] = self.PLAYED_JOB_SIG
        self._found_job = True

    def _extract_receipt_methods(self, artifacts: dict, class_data: str) -> None:
        methods = list(self.METHOD_RE.finditer(class_data))

        receipt_types = [method for method in methods
                         if 'public' in method.group('modifiers')
                         and self.RECEIPT_TYPE_SIG_RE.match(method.group('sig'))
                         and self.READ_RE.search(method.group('body'))
                         and self.READ_SELF_RE.search(method.group('body'))]
        if len(receipt_types) != 1:
            return
        receipt_type = receipt_types[0]

        # The helpers are whichever private predicates the receipt-type method
        # consults; naming them outright would hard-code an obfuscated name.
        helpers = [method.group('name') for method in methods
                   if 'private' in method.group('modifiers')
                   and self.PLAYED_GATE_SIG_RE.match(method.group('sig'))
                   and self._calls(receipt_type.group('body'), method.group('name'))]
        if len(helpers) < self.MIN_SHARED_HELPERS:
            return

        played_gates = [method for method in methods
                        if 'public' in method.group('modifiers')
                        and self.PLAYED_GATE_SIG_RE.match(method.group('sig'))
                        and sum(1 for helper in helpers
                                if self._calls(method.group('body'), helper))
                        >= self.MIN_SHARED_HELPERS]
        if len(played_gates) != 1:
            return
        played_gate = played_gates[0]

        class_match = CLASS_RE.search(class_data)
        if class_match is None:
            return

        artifacts['READ_RECEIPT_UTILS_CLASS_NAME'] = class_match.groupdict().get('name').replace('/', '.')
        artifacts['RECEIPT_TYPE_METHOD_NAME'] = receipt_type.group('name')
        artifacts['RECEIPT_TYPE_METHOD_SIG'] = receipt_type.group('sig')
        artifacts['PLAYED_RECEIPT_GATE_METHOD_NAME'] = played_gate.group('name')
        artifacts['PLAYED_RECEIPT_GATE_METHOD_SIG'] = played_gate.group('sig')
        self._found_utils = True

File: smali_generator/app/src/main/java/com/smali_generator/patches/ReadReceipts.java
package com.smali_generator.patches;

import android.util.Log;

import com.arthooks.ArtHooks;

import com.smali_generator.Hook;
import com.smali_generator.HookCategory;
import com.smali_generator.db.PatchDb;
import com.smali_generator.ui.ChatPickerActivity;

import java.lang.reflect.Executable;
import java.lang.reflect.Method;
import java.util.Locale;


/**
 * Stops read receipts from reaching the sender.
 *
 * WhatsApp never decides whether to send a receipt, only which kind: a
 * {@code read} receipt goes to the sender and turns their ticks blue, while a
 * {@code read-self} receipt goes only to your own linked devices, so they mark
 * the chat read without telling anyone. Voice notes have the same pair,
 * {@code played} and {@code played-self}. The app already sends the self
 * variants -- for your own status views, for chats it considers spam, and for
 * everyone when its own Read receipts setting is off -- so this hook does not
 * invent a state: it takes a path the app takes for itself.
 *
 * Both decisions run through the class its log strings call ReadReceiptUtils.
 * The read one has a method of its own, which is simply forced to the self
 * variant. The played one does not, and is the reason for the third hook
 * below.
 *
 * Doing it here rather than at the app's own privacy setting is what keeps the
 * asymmetry worth having: flipping that setting also stops *you* from seeing
 * other people's blue ticks, and this does not.
 *
 * The played gate covers the blue microphone on voice notes. It is included
 * because the two are one setting in WhatsApp's own UI, so leaving it out
 * would leak the very thing the hook is for -- grey ticks on a chat, and then
 * a blue mic the moment a voice note is opened.
 *
 * That gate is not a played-receipt method, though: it is the app's general
 * "are read receipts on for this chat", with twelve callers on 2.26.36.71.
 * One is the job that sends the played receipt; another is
 * MessageStatusUpdateReceiptFactory, which reads it when a receipt arrives
 * from the other side and, on a "no", rewrites the incoming read or played
 * status down to delivered -- WhatsApp's own reciprocity rule. Answering no
 * everywhere therefore turns that rule on the user and hides other people's
 * blue ticks, which is the one thing this hook promises not to do. So the
 * answer is scoped: {@link #played_job_hook} marks the job while it runs, and
 * the gate lies only to the single call made inside it.
 *
 * Which chats it applies to is a {@link Scope}, chosen on {@link
 * ChatPickerActivity}. Both hooked methods take the chat's Jid, so the scope
 * is consulted per receipt rather than per process -- unlike the switch that
 * installs the hook at all, changing the scope takes effect immediately.
 */
public class ReadReceipts implements Hook {

    private static final String TAG = "PATCH";

    /** The receipt that syncs to your own devices and goes nowhere else. */
    private static final String READ_SELF = "read-self";

    /** Key for this hook's chat list, and for its scope setting. */
    public static final String FEATURE = "read_receipts";
    private static final String SCOPE_KEY = "read_receipts_scope";

    /**
     * Which chats receipts are held back from.
     *
     * One list covers both directions people ask for: {@code ONLY_LISTED} is a
     * blocklist and {@code ALL_EXCEPT_LISTED} an allowlist over the same set of
     * chats, so switching between them keeps the picks.
     */
    public enum Scope {
        EVERYONE("Every chat", "Nobody sees your read receipts."),
        ONLY_LISTED("Only the chats I pick", "Hidden from the chats below. Everyone else sees them as usual."),
        ALL_EXCEPT_LISTED("Every chat except the ones I pick", "Hidden from everyone but the chats below.");

        private final String title;
        private final String caption;

        Scope(String title, String caption) {
            this.title = title;
            this.caption = caption;
        }

        public String title() {
            return title;
        }

        public String caption() {
            return caption;
        }
    }

    /** Resolved once at load; the Jid class is the host app's, so it is found by reflection. */
    private static volatile Method rawJidMethod;

    public static Scope scope() {
        String stored = PatchDb.getString(SCOPE_KEY, Scope.EVERYONE.name());
        try {
            return Scope.valueOf(stored);
        } catch (IllegalArgumentException e) {
            // A scope written by a newer build, or a corrupted row. Holding
            // everything back is the reading that cannot leak a receipt.
            Log.e(TAG, "ReadReceipts: unknown scope " + stored + ", falling back to every chat");
            return Scope.EVERYONE;
        }
    }

    public static void setScope(Scope scope) {
        PatchDb.setString(SCOPE_KEY, scope.name());
    }

    /** Volatile so "once" holds across the threads receipts are sent from. */
    private static volatile boolean logged_read;
    private static volatile boolean logged_played;

    /**
     * Set while this thread is inside the played-receipt job.
     *
     * The gate is shared, so this is what separates the one caller that is
     * deciding what to *send* from the many that are deciding what to *show*.
     */
    private static final ThreadLocal<Boolean> in_played_job = new ThreadLocal<>();

    /**
     * Why one chat's receipts were, or were not, held back.
     *
     * Carries no chat identifier, and neither does anything logged from it:
     * this runs for every message read, and a log line naming the chat would
     * be a record of who the user talks to and when -- the thing the hook
     * exists to keep from leaking.
     */
    private static final class Decision {
        final boolean suppress;
        private final Scope scope;
        private final String chat;

        private Decision(boolean suppress, Scope scope, String chat) {
            this.suppress = suppress;
            this.scope = scope;
            this.chat = chat;
        }

        /** Nothing about the chat was looked at, so the line must not claim otherwise. */
        static Decision everyChat(Scope scope) {
            return new Decision(true, scope, null);
        }

        static Decision unidentified(Scope scope) {
            return new Decision(true, scope, "chat unidentified");
        }

        static Decision forChat(Scope scope, boolean picked) {
            return new Decision(scope == Scope.ONLY_LISTED ? picked : !picked, scope,
                    picked ? "chat is picked" : "chat is not picked");
        }

        @Override
        public String toString() {
            return "scope=" + scope + (chat == null ? "" : ", " + chat)
                    + (suppress ? ", held back" : ", sent as usual");
        }
    }

    /**
     * Whether this chat's receipts should be held back.
     *
     * One decision for both receipt kinds. A chat that cannot be identified is
     * held back whatever the scope says: the two ways to be wrong are not
     * equal, since a receipt withheld by mistake is invisible and one sent by
     * mistake cannot be taken back.
     */
    private static Decision decide(Object jid) {
        Scope scope = scope();
        if (scope == Scope.EVERYONE) {
            return Decision.everyChat(scope);
        }
        String raw = raw_jid(jid);
        if (raw == null) {
            return Decision.unidentified(scope);
        }
        return Decision.forChat(scope, PatchDb.isChatSelected(FEATURE, raw));
    }

    /**
     * The chat's jid as text, which is what a selection is keyed on.
     *
     * Deliberately not toString(): that returns the Jid's obfuscated form,
     * which is what the app logs rather than what identifies a chat.
     */
    private static String raw_jid(Object jid) {
        Method method = rawJidMethod;
        if (method == null || jid == null) {
            return null;
        }
        try {
            Object raw = method.invoke(jid);
            return raw instanceof String ? (String) raw : null;
        } catch (Throwable t) {
            return null;
        }
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     * silently answer for the original. ArtHooks rewrites its entry point. */
    static native String receipt_type_backup(Object thiz, Object jid, boolean force_read_self);

    static String receipt_type_hook(Object thiz, Object jid, boolean force_read_self) {
        Decision decision = decide(jid);
        // Always asked, even when the answer is discarded: it is what the app
        // would have done, so a chat that is not held back keeps WhatsApp's own
        // rules rather than a second implementation of them. Calling through is
        // safe and cheap -- the original only reads settings and chat state.
        String original = receipt_type_backup(thiz, jid, force_read_self);
        if (!logged_read) {
            logged_read = true;
            Log.i(TAG, "ReadReceipts: first read receipt: " + decision
                    + ", app wanted \"" + original + "\"");
        }
        return decision.suppress ? READ_SELF : original;
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     * silently answer for the original. ArtHooks rewrites its entry point. */
    static native boolean played_gate_backup(Object thiz, Object jid);

    static boolean played_gate_hook(Object thiz, Object jid) {
        boolean original = played_gate_backup(thiz, jid);
        if (!Boolean.TRUE.equals(in_played_job.get())) {
            // Every other caller is asking so it can decide what to display,
            // and gets the app's own answer untouched. Note this returns
            // before decide(), so the shared path costs one ThreadLocal read.
            return original;
        }
        // Consumed rather than merely read: the job asks once, early, and
        // whatever it calls afterwards on this thread is no longer the send
        // decision. This narrows the lie to exactly that one question.
        in_played_job.remove();
        Decision decision = decide(jid);
        if (!logged_played) {
            logged_played = true;
            Log.i(TAG, "ReadReceipts: first played receipt: " + decision
                    + ", app wanted \"" + (original ? "played" : "played-self") + "\"");
        }
        // false picks "played-self", the voice-note counterpart of read-self.
        return decision.suppress ? false : original;
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     * silently answer for the original. ArtHooks rewrites its entry point. */
    static native void played_job_backup(Object thiz);

    /**
     * Brackets one run of the played-receipt job, so the gate can tell that
     * run's question apart from the same question asked anywhere else.
     */
    static void played_job_hook(Object thiz) {
        in_played_job.set(Boolean.TRUE);
        try {
            played_job_backup(thiz);
        } finally {
            // Cleared even when the job throws -- it is allowed to, the queue
            // retries it. A flag left behind would follow the next piece of
            // work onto this pooled thread and suppress a receipt nobody asked
            // about, or worse, clamp an incoming one.
            in_played_job.remove();
        }
    }

    public String id() {
        return "read_receipts";
    }

    public String title() {
        return "Hide when you've read a message";
    }

    public String description() {
        return "Nobody sees your blue ticks, or the blue microphone on voice notes. You still see theirs.";
    }

    public HookCategory category() {
        return HookCategory.PRIVACY;
    }

    public String configSummary() {
        Scope scope = scope();
        if (scope == Scope.EVERYONE) {
            return "Applies to every chat";
        }
        int count = PatchDb.selectedChats(FEATURE).size();
        String chats = count == 1 ? "1 chat" : count + " chats";
        return String.format(Locale.US, scope == Scope.ONLY_LISTED
                ? "Applies to %s" : "Applies to every chat except %s", chats);
    }

    public Class<?> configScreen() {
        return ChatPickerActivity.class;
    }

    public void load() {
        try {
            rawJidMethod = Class.forName("{{JID_CLASS_NAME}}")
                    .getMethod("{{JID_RAW_STRING_METHOD_NAME}}");
            Class<?> read_receipt_utils = Class.forName("{{READ_RECEIPT_UTILS_CLASS_NAME}}");

            Executable receipt_type = ArtHooks.find_function(read_receipt_utils,
                    "{{RECEIPT_TYPE_METHOD_NAME}}", "{{RECEIPT_TYPE_METHOD_SIG}}");
            Method receipt_type_replacement = ReadReceipts.class.getDeclaredMethod(
                    "receipt_type_hook", Object.class, Object.class, boolean.class);
            Method receipt_type_original = ReadReceipts.class.getDeclaredMethod(
                    "receipt_type_backup", Object.class, Object.class, boolean.class);
            boolean read_hooked = ArtHooks.hook_function(
                    receipt_type, receipt_type_replacement, receipt_type_original);

            Executable played_gate = ArtHooks.find_function(read_receipt_utils,
                    "{{PLAYED_RECEIPT_GATE_METHOD_NAME}}", "{{PLAYED_RECEIPT_GATE_METHOD_SIG}}");
            Method played_gate_replacement = ReadReceipts.class.getDeclaredMethod(
                    "played_gate_hook", Object.class, Object.class);
            Method played_gate_original = ReadReceipts.class.getDeclaredMethod(
                    "played_gate_backup", Object.class, Object.class);
            boolean played_hooked = ArtHooks.hook_function(
                    played_gate, played_gate_replacement, played_gate_original);

            Class<?> played_job = Class.forName("{{PLAYED_RECEIPT_JOB_CLASS_NAME}}");
            Executable played_run = ArtHooks.find_function(played_job,
                    "{{PLAYED_RECEIPT_JOB_METHOD_NAME}}", "{{PLAYED_RECEIPT_JOB_METHOD_SIG}}");
            Method played_job_replacement = ReadReceipts.class.getDeclaredMethod(
                    "played_job_hook", Object.class);
            Method played_job_original = ReadReceipts.class.getDeclaredMethod(
                    "played_job_backup", Object.class);
            boolean job_hooked = ArtHooks.hook_function(
                    played_run, played_job_replacement, played_job_original);

            // job= is worth reading: without it the gate never fires, so the
            // blue microphone leaks while everything else looks installed.
            Log.i(TAG, "ReadReceipts: Patch loaded on " + read_receipt_utils.getName()
                    + ", read=" + read_hooked + " played=" + played_hooked + " job=" + job_hooked
                    + ", scope=" + scope() + " chats=" + PatchDb.selectedChats(FEATURE).size());
        } catch (Throwable t) {
            Log.e(TAG, "ReadReceipts: Error: " + t);
        }
    }

    public void unload() {
        Log.i(TAG, "ReadReceipts: Patch unloaded");
    }
}
